crates/neon/src/types_impl/extract/container.rs

```diff
@@ -0,0 +1,150 @@
+use std::{
+    cell::{Ref, RefCell, RefMut},
+    rc::Rc,
+    sync::Arc,
+};
+
+use crate::{
+    context::Cx,
+    handle::Handle,
+    result::{JsResult, NeonResult},
+    types::{
+        extract::{TryFromJs, TryIntoJs},
+        JsBox, JsValue,
+    },
+};
+
+use super::error::{RefCellError, RustTypeExpected};
+
+pub trait Container {
+    fn container_name() -> &'static str;
+}
+
+impl<T> Container for RefCell<T> {
+    fn container_name() -> &'static str {
+        "std::cell::RefCell"
+    }
+}
+
+impl<T> Container for Rc<T> {
+    fn container_name() -> &'static str {
+        "std::rc::Rc"
+    }
+}
+
+impl<T> Container for Arc<T> {
+    fn container_name() -> &'static str {
+        "std::sync::Arc"
+    }
+}
+
+impl<'cx, T: 'static> TryFromJs<'cx> for &'cx RefCell<T> {
+    type Error = RustTypeExpected<RefCell<T>>;
```

Comment:
What do you think of using `TypeExpected<JsBox<T>>` to reduce the API surface area a bit? Since these types are still `JsBox` under the hood, I think it's okay to expose the error as that.

Reply:
That may be a good idea also so people can understand what the actual runtime representation is. (I think it would be `TypeExpected<JsBox<RefCell<T>>>`, right?)

Reply:
@kjvalencik A related design call I'd love your input on: right now the dynamic borrow errors are explicitly handled with `try_borrow` and `try_borrow_mut` and turned into a Neon-specific error type called `RefCellError`. An alternative is we could just let them panic and keep the API surface to just `TypeExpected`. Do you have a preference?

Reply:
I also dislike the fact that `RefCellError::WrongType` overlaps with `TypeExpected`.

Reply:
Yeah I really don't see much tangible benefit to the extra types. The panic message is basically the same. I'll remove it for now and if you disagree we can talk about putting the type back.

Reply:
Yes, `TypeExpected<JsBox<RefCell<T>>>` would be the type I would expect. I don't like letting it panic. We'll catch the unwind and convert it to an exception, but it might not be clear.

I think it's okay to let throw an exception. If you don't want this behavior, the author can go to `&'cx RefCell` first and then borrow. If they are going directly to `Ref` or `RefMut`, they are already opting into some shorthand syntax.

However, I think it's better for the `try_from_js` to call the `try_` methods on `RefCell` and on error, manually throw with `cx.throw_error(..)` and a clear error message.

Reply:
Oh I think I see what you're suggesting: instead of creating a new Error type, just explicitly throw an error with a custom message directly in the `TryFromJs` implementation. That's reasonable. I'm not sure it makes a huge difference but I'm OK with it.

Reply:
OK I pushed this middle ground -- it uses `try_borrow()` and `try_borrow_mut()` but explicitly calls `cx.throw_error()` instead of creating a custom type, so it can use `TypeExpected` for the other case. This feels like a good choice to me but see what you think.